src/Raven.Server/Documents/Indexes/Persistence/Lucene/OutputReduceIndexWriteOperation.cs

```diff
@@ -65,55 +64,78 @@ public override void DeleteReduceResult(LazyStringValue reduceKeyHash, IndexingS
             _outputReduceToCollectionCommand.DeleteReduce(reduceKeyHash);
         }
 
+        public override void Dispose()
+        {
+            base.Dispose();
+
+            _outputReduceToCollectionCommand.Dispose();
+        }
+
         public class OutputReduceToCollectionCommand : TransactionOperationsMerger.MergedTransactionCommand, IDisposable
         {
+            private const string MultipleOutputsForSameReduceKeyHashSeparator = "/";
+
             private readonly DocumentDatabase _database;
             private readonly string _outputReduceToCollection;
             private readonly MapReduceIndex _index;
-            private readonly List<OutputReduceDocument> _reduceDocuments = new List<OutputReduceDocument>();
-            private readonly JsonOperationContext _jsonContext;
-            private readonly List<IDisposable> _disposables = new List<IDisposable>();
-            public OutputReduceToCollectionCommand(DocumentDatabase database, string outputReduceToCollection, MapReduceIndex index)
+            private readonly List<string> _reduceKeyHashesToDelete = new List<string>();
+
+            private readonly Dictionary<string, List<BlittableJsonReaderObject>> _reduceDocuments =
+                new Dictionary<string, List<BlittableJsonReaderObject>>();
+            private readonly JsonOperationContext _indexContext;
+
+            public OutputReduceToCollectionCommand(DocumentDatabase database, string outputReduceToCollection, MapReduceIndex index, JsonOperationContext context)
             {
                 _database = database;
                 _outputReduceToCollection = outputReduceToCollection;
                 _index = index;
-                _jsonContext = JsonOperationContext.ShortTermSingleUse();
-            }
-
-            private class OutputReduceDocument
-            {
-                public bool IsDelete;
-                public string Key;
-                public BlittableJsonReaderObject Document;
+                _indexContext = context;
             }
 
             public override int Execute(DocumentsOperationContext context)
             {
-                foreach (var reduceDocument in _reduceDocuments)
+                foreach (var reduceKeyHash in _reduceKeyHashesToDelete)
                 {
-                    var key = reduceDocument.Key;
-                    if (reduceDocument.IsDelete)
+                    var deleteResult = _database.DocumentsStorage.Delete(context, GetOutputDocumentKey(reduceKeyHash), null);
```

Comment:
What happens if we have a hash collision from _another_ batch on unrelated doc?

Reply:
If we have hash collision then we're guaranteed that map results will be put into _the same_ reduce tree. Here we output docs per reduce tree so we're good.

Another batch of docs with collision would end up into putting them to the same reduce tree - we always output _all_ documents from a given reduce tree